src/System.Private.CoreLib/src/System/Threading/ThreadStore.cs

```diff
@@ -0,0 +1,57 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime;
+
+namespace System.Threading
+{
+    internal static class ThreadStore
+    {
+        private static readonly ManualResetEvent _allDone = new ManualResetEvent(false);
+
+        // main thread started by runtime and we have no chance (?) to increment it elsewhere
+        private static int _foregroundRunningCount = 1;
+
+        internal static void IncrementRunningForeground() 
+        {
+            Interlocked.Increment(ref _foregroundRunningCount);
+        }
+
+        internal static void DecrementRunningForeground() 
+        {
+            if (Interlocked.Decrement(ref _foregroundRunningCount) == 0) 
+            {
+                _allDone.Set();
```

Comment:
I think this can be set only once we are in the shutdown. For example, the program may set all threads to background (even the Main thread) and then set some of the threads back to foreground. I believe that the expected behavior is to wait for the foreground threads in this case.

Reply:
I agree with your logic and brainstormed in that way while implementing this patch. But this is how it works in CoreCLR and I decided to keep things consistent:

https://github.com/dotnet/runtime/blob/671f99b4f00e6c47926b09e3dcf899f154ca5d31/src/coreclr/src/vm/threads.cpp#L4607

Also note comments for `ThreadStore::CheckForEEShutdown()`
> // Whenever one of the components of OtherThreadsComplete() has changed in the
// correct direction, see whether we can now shutdown the EE because only background
// threads are running.

However if you thing that this is desired behavior I can keep it hacking.

Reply:
Ah ok. We can just keep doing what CoreCLR does in that case.